src/helpers.cc

```diff
@@ -0,0 +1,71 @@
+#include <string>
+
+#include <google/protobuf/stubs/strutil.h>
+
+#include <helpers.hh>
+#include <primitive_field.hh>
+#include <field_base.hh>
+
+namespace chapel {
+  
+  string StripDotProto(const std::string& proto_file) {
+    int lastindex = proto_file.find_last_of(".");
+    return proto_file.substr(0, lastindex);
+  }
+  
+  string UnderscoresToCamelCase(const string& input) {
+    string result;
+    bool cap_next_letter = false;
+    for (int i = 0; i < input.size(); i++) {
+      if ('a' <= input[i] && input[i] <= 'z') {
+        if (cap_next_letter) {
+          result += input[i] + ('A' - 'a');
+        } else {
+          result += input[i];
+        }
+        cap_next_letter = false;
+      } else if ('A' <= input[i] && input[i] <= 'Z') {
+        if (cap_next_letter) {
+          result += input[i];
+        } else {
+          result += input[i] + ('a' - 'A');
+        }
+        cap_next_letter = false;
+      } else if ('0' <= input[i] && input[i] <= '9') {
+        result += input[i];
+        cap_next_letter = true;
+      } else {
+        cap_next_letter = true;
+      }
+    }
+    return result;
+  }
+  
+  string GetFileNameBase(const FileDescriptor* descriptor) {
+      string proto_file = descriptor->name();
+      int lastslash = proto_file.find_last_of("/");
+      string base = proto_file.substr(lastslash + 1);
+      return UnderscoresToCamelCase(StripDotProto(base));
+  }
+
+  string GetOutputFile(const FileDescriptor* descriptor, string* error) {
+    string relative_filename = GetFileNameBase(descriptor);
```

Comment:
This seems to always result in the .chpl file being placed in the current directory?  I may not have followed the call chain back far enough, though

Reply:
No, the placement of the `.chpl` file is decided by the `GeneratorContext` object which deals with the output path passed to the `protoc` compiler as

`protoc --chpl_out=/output/dir/ a.proto`

This is an internal implementation in `google/protobuf/compiler/code_generator.cc`.